### PR TITLE
Fix status.address.url for networking layer other than istio

### LINKED ISSUE #2856: The image classification inference graph example does not work with Kourier
/kind bug

**What steps did you take and what happened:**
The example in https://kserve.github.io/website/0.10/modelserving/inference_graph/image_pipeline/#deploy-the-inferenceservices does not work with kourier. It might be related to https://github.com/kserve/kserve/pull/2830

When the pipeline is called, the pipeline container tries to call `http://cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local/v1/models/cat-dog-classifier:predict`. But the service is called 
```
k get svc -n kserve-demo cat-dog-classifier-predictor
NAME                           TYPE           CLUSTER-IP   EXTERNAL-IP                                         PORT(S)   AGE
cat-dog-classifier-predictor   ExternalName   <none>       kourier-internal.kourier-system.svc.cluster.local   80/TCP    7m44s
```
so without `-default`.

Full error log:
```
user-container {"level":"info","ts":1682501060.9219692,"logger":"InferenceGraphRouter","msg":"Entering callService","url":"http://cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local/v1/models/cat-dog-classifier:predict"}
user-container {"level":"error","ts":1682501060.9319263,"logger":"InferenceGraphRouter","msg":"An error has occurred from service","service":"http://cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local/v1/models/cat-dog-classifier:predict","error":"Post \"http://cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local/v1/models/cat-dog-classifier:predict\": dial tcp: lookup cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local on 10.96.0.10:53: no such host","stacktrace":"main.callService\n\t/go/src/github.com/kserve/kserve/cmd/router/main.go:58\nmain.executeStep\n\t/go/src/github.com/kserve/kserve/cmd/router/main.go:182\nmain.routeStep\n\t/go/src/github.com/kserve/kserve/cmd/router/main.go:167\nmain.graphHandler\n\t/go/src/github.com/kserve/kserve/cmd/router/main.go:189\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\nnet/http.(*ServeMux).ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2462\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2916\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1966"}
user-container {"level":"info","ts":1682501060.9320958,"logger":"InferenceGraphRouter","msg":"elapsed time","step":"http://cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local/v1/models/cat-dog-classifier:predict","time":0.010212725}
user-container {"level":"info","ts":1682501060.9321399,"logger":"InferenceGraphRouter","msg":"elapsed time","node":"root","time":0.010267933}
user-container {"level":"error","ts":1682501060.9321415,"logger":"InferenceGraphRouter","msg":"failed to process request","error":"Post \"http://cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local/v1/models/cat-dog-classifier:predict\": dial tcp: lookup cat-dog-classifier-predictor-default.kserve-demo.svc.cluster.local on 10.96.0.10:53: no such host","stacktrace":"main.graphHandler\n\t/go/src/github.com/kserve/kserve/cmd/router/main.go:190\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2084\nnet/http.(*ServeMux).ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2462\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2916\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1966"}
```

**What did you expect to happen:**
The example should work.

**What's the InferenceService yaml:**
Yamls from https://kserve.github.io/website/0.10/modelserving/inference_graph/image_pipeline/#deploy-the-inferenceservices.

**Anything else you would like to add:**
-

**Environment:**

- Istio Version: -
- Knative Version: 1.9
- KServe Version: 0.10.1
- Kubeflow version: -
- Cloud Environment: -
- Minikube/Kind version: Kind 0.18.0
- Kubernetes version: 1.26
- OS: MacOS

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2856 

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go [hunk 1]
```diff
@@ -542,9 +543,9 @@ func routeSemanticEquals(desired, existing *v1alpha3.VirtualService) bool {
 func getHostPrefix(isvc *v1beta1.InferenceService, disableIstioVirtualHost bool) string {
 	if disableIstioVirtualHost == true {
 		if isvc.Spec.Transformer != nil {
-			return constants.DefaultTransformerServiceName(isvc.Name)
+			return constants.TransformerServiceName(isvc.Name)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: We need to add a check if the transformer or predictor exists with the `default` suffix.

Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: we need an e2e test for disabling the top level istio virtual host

Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: I am confused if the original bug is for serverless mode or raw deployment mode

Comment (sivanantha321) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: This bug is only valid for serverless deployment with network layer other than istio

Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: But I think this problem is only for the master branch not in the release right ?

Comment (sivanantha321) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: yes, It is introduced when we removed default suffix

Comment (sivanantha321) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go]: @yuzisun Tested it with istio and kourier. It is working fine :)

File: test/e2e/common/utils.py [hunk 2]
```diff
@@ -194,16 +194,18 @@ def explain_response(service_name, input_json):
...
+        if service.status.load_balancer.ingress is None:
+            cluster_ip = service.spec.cluster_ip
         else:
-            cluster_ip = service.status.load_balancer.ingress[0].ip
-    return os.environ.get("KSERVE_INGRESS_HOST_PORT", cluster_ip)
+            if service.status.load_balancer.ingress[0].hostname:
+                cluster_ip = service.status.load_balancer.ingress[0].hostname
+            else:
+                cluster_ip = service.status.load_balancer.ingress[0].ip
+    return cluster_ip
```
Comment (yuzisun) [inline test/e2e/common/utils.py:208]: How is this different from the original logic?

File: test/overlays/knative/overlays/kourier/net_kourier_controller_deployment_patch.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    spec:
+      containers:
+        - name: controller
+          resources:
+            limits:
+              cpu: 100m
+              memory: 128Mi
+            requests:
+              cpu: 5m
+              memory: 32Mi
```
Comment (yuzisun) [inline test/overlays/knative/overlays/kourier/net_kourier_controller_deployment_patch.yaml:17]: I think we should consider migrating to use knative operator to configure these settings without using kustomize overlays

File: test/e2e/common/utils.py (hunk 2 above)
Comment (sivanantha321) [reply test/e2e/common/utils.py:208]: If we set the cluster ip in environment variable, then there is no need to create a client and check for the cluster ip. In previous logic even if we set the environmental variable it goes through the unnecessary steps like creating a client and check for the ingress ip and finally just returning what is set in the env variable.

File: test/overlays/knative/overlays/kourier/net_kourier_controller_deployment_patch.yaml (hunk 3 above)
Comment (sivanantha321) [reply test/overlays/knative/overlays/kourier/net_kourier_controller_deployment_patch.yaml:17]: Can we do it in a separate PR ?

Comment (yuzisun) [reply test/overlays/knative/overlays/kourier/net_kourier_controller_deployment_patch.yaml:17]: Created the issue https://github.com/kserve/kserve/issues/2982

**Pull request merged**